cranelift/codegen/src/egraph/stores.rs

```diff
@@ -0,0 +1,270 @@
+//! enter a block no longer dominated by that block, we clear the info
+//! to "unknown". (We could do a fixpoint analysis instead and resolve
+//! merges this way, but in practice when iterating over a
+//! structured-code CFG in RPO, our approach will do just as well.)
+//!
+//! We can't say for sure that the "last store" *did* actually write
+//! that state, but we know for sure that no instruction *later* than
+//! it (up to the current instruction) did. However, we can get a
+//! must-alias property from this: if at a given load or store, we
+//! look backward to the "last store", *AND* we find that it has
+//! exactly the same address expression and type, then we know that
+//! the current instruction's access *must* be to the same memory
+//! location.
+//!
+//! To get this must-alias property, we leverage the node
+//! hashconsing. We design the Eq/Hash (node identity relation
+//! definition) of the `Node` struct so that all loads with (i) the
+//! same "last store", and (ii) the same address expression, and (iii)
+//! the same opcode-and-offset, will deduplicate (the first will be
+//! computed, and the later ones will use the same value). Furthermore
+//! we have a rewrite rule (in opts/store_to_load.isle) that rewrites
+//! a load into the stored value of the last store *if* the last store
+//! has the same address expression and constant offset.
+//!
+//! This gives us two optimizations, "redundant load elimination" and
+//! "store-to-load forwarding".
+//!
+//! In theory we could also do *dead-store elimination*, where if a
+//! store overwrites a value earlier written by another store, *and*
+//! if no other load/store to the abstract state category occurred,
+//! *and* no other trapping instruction occurred (at which point we
+//! need an up-to-date memory state because post-trap-termination
+//! memory state can be observed), *and* we can prove the original
+//! store could not have trapped, then we can eliminate the original
+//! store. Because this is so complex, and the conditions for doing it
+//! correctly when post-trap state must be correct likely reduce the
+//! potential benefit, we don't yet do this.
+
+use crate::flowgraph::ControlFlowGraph;
+use crate::fx::{FxHashMap, FxHashSet};
+use crate::inst_predicates::has_memory_fence_semantics;
+use crate::ir::{Block, Function, Inst, Opcode};
+use cranelift_entity::EntityRef;
+use cranelift_entity::SecondaryMap;
+use smallvec::{smallvec, SmallVec};
+
+/// For a given program point, the vector of last-store instruction
+/// indices for each disjoint category of abstract state.
+#[derive(Clone, Copy, Debug, Default, PartialEq, Eq)]
+struct LastStores {
+    heap: MemoryState,
+    table: MemoryState,
+    vmctx: MemoryState,
+    other: MemoryState,
+}
+
+/// State of memory seen by a load.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum MemoryState {
+    Entry,
+    Store(Inst),
+    BeforeInst(Inst),
+    AfterInst(Inst),
+    Top,
```

Comment:
Right, it's not fully a meet-lattice (or a proper "meet" binary operator), because meet takes a location and generates a different "bottom" value for each location.

Or more properly, it is a family of meet semilattices, indexed on location; each location actually meets over a slightly different semilattice with its own private "bottom".

All that's a fancy and fairly useless way of saying that we want to not fall back to a generic "unknown, we can't conclude anything" state: we can actually still name the state (every program point has some symbolic state), we just don't have a better name for it or a proof that it is equal to any other state in the program.

So -- given that it's definitely not a meet-semilattice -- your suggestions are great actually and I've updated things (and added doc-comments). Thanks!